lisa/util/__init__.py

```diff
@@ -598,6 +601,108 @@ def is_valid_url(url: str, raise_error: bool = True) -> bool:
     return is_url
 
 
+def _raise_or_log_failure(log: Logger, raise_error: bool, failure_msg: str) -> bool:
+    if raise_error:
+        raise LisaException(failure_msg)
+    else:
+        log.info(failure_msg)
+        return False
+
+
+# big function to check the parts of a url
+# allow raising exceptions or log and return a bool
+# allows checks for:
+#   expected domains
+#   protocols (require https, sftp, etc)
+#   filenames (pattern matching)
+def check_url(
+    log: Logger,
+    source_url: str,
+    expected_package_name_pattern: Optional[Pattern[str]] = None,
+    allowed_protocols: Optional[List[str]] = None,
+    expected_domains: Optional[List[str]] = None,
```

Comment:
`expected_domain_pattern`. One pattern should be enough to validate all.

Reply:
I started with just a regex match but it was a pain with annoying edge cases, so I started writing this version. My goal is making it easier for the caller, this way you can pass a couple of domains, a couple of protocols, and a filename pattern (or just a string) and restrict future users from downloading whatever tarball from wherever later with little room for error.

Reply:
long story short I'm writing it for me so I don't mess it up next time I go to do this 😅

Reply:
You can test regexp at someplace like https://regex101.com. if it's string comparison, it will meet more problems actually. Like how to allow/disallow subdomains, or attack like "invalidmicrosoft.com"?

Reply:
Good point!

Reply:
Honestly, wish I hadn't even tried to handle this at this point.

Reply:
okedokey, moved to a pattern for domains,